main.go

```diff
@@ -0,0 +1,295 @@
+	"github.com/tidwall/resp"
+)
+
+var jwtPubKey *ecdsa.PublicKey
+
+type ZdbBackend struct {
+	Server redis.Conn
+}
+
+type Session struct {
+	Authenticated bool
+	Username      string
+	Backend       *ZdbBackend
+}
+
+type Sessions struct {
+	List    map[*resp.Conn]*Session
+	Backend *ZdbBackend
+}
+
+//
+// its you online
+//
+const (
+	iyoPubKey = `-----BEGIN PUBLIC KEY-----
+MHYwEAYHKoZIzj0CAQYFK4EEACIDYgAES5X8XrfKdx9gYayFITc89wad4usrk0n2
+7MjiGYvqalizeSWTHEpnd7oea9IQ8T5oJjMVH5cc0H5tFSKilFFeh//wngxIyny6
+6+Vq5t5B0V0Ehy01+2ceEon2Y0XDkIKv
+-----END PUBLIC KEY-----`
+)
+
+func cryptoInit() {
+	var err error
+
+	jwtPubKey, err = jwt.ParseECPublicKeyFromPEM([]byte(iyoPubKey))
+	if err != nil {
+		log.Panicf("failed to parse pub key:%v\n", err)
+	}
+}
+
+//
+// list of sessions
+//
+func NewSessions(backend *ZdbBackend) *Sessions {
+	return &Sessions{
+		List:    make(map[*resp.Conn]*Session),
+		Backend: backend,
+	}
+}
+
+// single session
+func (ss *Sessions) NewSession() *Session {
+	return &Session{
+		Authenticated: false,
+		Username:      "",
+		Backend:       ss.Backend,
+	}
+}
+
+// create a new Session based on a new connection
+func (ss *Sessions) SessionAccept(conn *resp.Conn) bool {
+	log.Printf("Creating a new session for [%v]\n", conn.RemoteAddr)
+	ss.List[conn] = ss.NewSession()
+	return true
+}
+
+// verify a jwt token is valid
+func verifyToken(tokenStr string) (bool, string, error) {
+	token, err := jwt.Parse(tokenStr, func(token *jwt.Token) (interface{}, error) {
+		if token.Method != jwt.SigningMethodES384 {
+			return nil, fmt.Errorf("Unexpected signing method: %v", token.Header["alg"])
+		}
+		return jwtPubKey, nil
+	})
+
+	if err != nil {
+		return false, "", err
+	}
+
+	if claims, ok := token.Claims.(jwt.MapClaims); ok && token.Valid {
+		log.Printf("Token validated: %v", claims["username"])
+		return true, claims["username"].(string), nil
+	}
+
+	return false, "", nil
+}
+
+//
+// AUTH
+//
+func (ss *Sessions) Authenticate(conn *resp.Conn, args []resp.Value) bool {
+	// retreive session from connection
+	session := ss.List[conn]
+	return session.Authenticate(conn, args)
+}
+
+func (s *Session) Authenticate(conn *resp.Conn, args []resp.Value) bool {
+	if len(args) != 2 {
+		conn.WriteError(errors.New("Wrong number of arguments for 'AUTH' command"))
+		return true
+	}
+
+	valid, username, err := verifyToken(args[1].String())
+	if !valid {
+		log.Printf("Authentication failed: %v\n", err)
+		conn.WriteError(errors.New("Access denied"))
+
+		s.Authenticated = false
+		return true
+	}
+
+	s.Username = username
+	s.Authenticated = true
+
+	conn.WriteSimpleString("OK - " + s.Username)
+	return true
+}
+
+//
+// INFO
+//
+func (ss *Sessions) Info(conn *resp.Conn, args []resp.Value) bool {
+	conn.WriteString("Gateway for 0-db (zdb) compatible clients\n")
+	return true
+}
+
+//
+// EXISTS
+//
+func (ss *Sessions) Exists(conn *resp.Conn, args []resp.Value) bool {
+	// retreive session from connection
+	session := ss.List[conn]
```

Comment:
you should probably handle the case where the connection doesn't exist in the map, otherwise this will panic